core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala

```diff
@@ -0,0 +1,116 @@
+) extends GraphStageLogic(shape) with PromiseControl with MessageBuilder[K, V, Msg] {
+  var consumer: ActorRef = _
+  var self: StageActor = _
+  var tps = Set.empty[TopicPartition]
+  var buffer: Iterator[ConsumerRecord[K, V]] = Iterator.empty
+  var requested = false
+  implicit lazy val ec = ActorMaterializer.downcast(materializer).executionContext
+
+  override def preStart(): Unit = {
+    super.preStart()
+    setKeepGoing(true)
+
+    consumer = {
+      val as = ActorMaterializer.downcast(materializer).system.asInstanceOf[ExtendedActorSystem]
+      val name = s"kafka-consumer-${KafkaConsumerActor.nextNumber()}"
+      as.systemActorOf(KafkaConsumerActor.props(settings), name)
+    }
+
+    subscription match {
+      case TopicSubscription(topics) =>
+        consumer ! KafkaConsumerActor.Subscribe(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case TopicSubscriptionPattern(topics) =>
+        consumer ! KafkaConsumerActor.SubscribePattern(topics, KafkaConsumerActor.rebalanceListener(partitionAssignedCB.invoke, partitionRevokedCB.invoke))
+      case Assignment(topics) =>
+        consumer ! KafkaConsumerActor.Assign(topics)
+        tps ++= topics
+      case AssignmentWithOffset(topics) =>
+        consumer ! KafkaConsumerActor.AssignWithOffset(topics)
+        tps ++= topics.keySet
+    }
+
+    self = getStageActor {
+      case (_, msg: KafkaConsumerActor.Messages[K, V]) =>
+        requested = false
+        // do not use simple ++ because of https://issues.scala-lang.org/browse/SI-9766
+        if (buffer.hasNext) {
+          buffer = buffer ++ msg.messages
+        }
+        else {
+          buffer = msg.messages
+        }
+        pump()
+      case (_, Terminated(ref)) if ref == consumer =>
+        failStage(new Exception("Consumer actor terminated"))
+    }
+    self.watch(consumer)
+  }
+
+  val partitionAssignedCB = getAsyncCallback[Iterable[TopicPartition]] { newTps =>
+    tps ++= newTps
+    pump()
+  }
+  val partitionRevokedCB = getAsyncCallback[Iterable[TopicPartition]] { newTps =>
+    tps --= newTps
+    pump()
+  }
+
+  @tailrec
+  private def pump(): Unit = {
+    if (isAvailable(shape.out)) {
+      if (buffer.hasNext) {
+        val msg = buffer.next()
+        push(shape.out, createMessage(msg))
+        pump()
+      }
+      else if (!requested && tps.nonEmpty) {
+        requested = true
+        consumer.tell(RequestMessages(tps), self.ref)
+      }
+    }
+  }
+
+  setHandler(shape.out, new OutHandler {
+    override def onPull(): Unit = {
+      pump()
+    }
+
+    override def onDownstreamFinish(): Unit = {
+      performShutdown()
+    }
+  })
+
+  override def performShutdown() = {
+    if (!isClosed(shape.out)) {
+      complete(shape.out)
+    }
+    self.become {
+      case (_, Terminated(ref)) if ref == consumer =>
+        shutdowned()
+        completeStage()
+    }
+    consumer ! KafkaConsumerActor.Stop
```

Comment:
I think it might be enough to send an extra Stop message is postStop in case the above block has not been called. It is an abort case anyway, so it is fine to be best effort.